lib/src/metta/runner/modules/catalog.rs

```diff
@@ -0,0 +1,568 @@
+// has a good reason to alias a module, it will now be able to do it.  I will implement this shortly,
+// which will involve deleting this comment altogether.  So this update is part of the "historical record"
+//
+pub trait ModuleCatalog: std::fmt::Debug + Send + Sync {
+    /// Returns the [ModuleDescriptor] for every module in the `ModuleCatalog` with the specified name
+    fn lookup(&self, name: &str) -> Vec<ModuleDescriptor>;
+
+    //TODO: Add this function when I add module versioning
+    // /// Returns the [ModuleDescriptor] for every module in the `ModuleCatalog` with the specified name
+    // /// matching the version requirements
+    // fn lookup_within_version_range(name: &str, version_range: ) -> Vec<ModuleDescriptor>;
+    //TODO: provide default implementation
+
+    //TODO: Add this function when I add module versioning
+    // /// Returns the [ModuleDescriptor] for the newest module in the `ModuleCatalog`, that falls within
+    // /// the specified version range, or `None` if no module exists
+    // fn lookup_newest_within_version_range(name: &str, version_range: ) -> Option<ModuleDescriptor>;
+    //TODO: provide default implementation
+
+    /// Returns a [ModuleLoader] for the specified module from the `ModuleCatalog`
+    fn get_loader(&self, descriptor: &ModuleDescriptor) -> Result<Box<dyn ModuleLoader>, String>;
+}
+
+/// The object responsible for locating and selecting dependency modules for each [MettaMod]
+///
+/// This structure is conceptually analogous to the a `Cargo.toml` file for a given module.
+#[derive(Clone, Debug, Default)]
+//TODO: Use serde to deserialize a PkgInfo from an expression atom
+pub struct PkgInfo {
+
+    /// The public name of the module.  Should be composed of alpha-numeric characters with '-' and '_'
+    /// characters allowed.  Must not contain any other punctuation. 
+    pub name: String,
+
+    //TODO: version field, to indicate the version of this module
+
+    /// If `strict == true` then a dependency must be declared in the `PkgInfo`, otherwise a permissive
+    /// version requirement will be assumed for any modules that are not explicitly declared
+    pub strict: bool,
+
+    /// Entries mapping module names to requirements for each dependency sub-module
+    ///
+    /// A Duplicate entry for a given sub-module in the deps list is an error.
+    pub deps: HashMap<String, DepEntry>,
+}
+
+/// A single entry in a [PkgInfo]'s dependencies, specifying the properties of a module that will satisfy a dependency
+#[derive(Clone, Debug, Default)]
+pub struct DepEntry {
+    /// Indicates that the dependency module should be loaded from a specific FS path
+    ///
+    /// If the fs_path is specified, the other pkg_info attributes will be ignored.
+    //QUESTION: We need a MeTTa "style guide" for these field names, since they are effective going
+    // to be part of the API, because a PkgInfo will be deserialized from atoms
+    pub fs_path: Option<PathBuf>
```

Comment:
This is related to the `atom-serde`; and before I implement `atom-serde` I'll probably use Hjson or TOML (which do you prefer?) to encode a pkg-info file for a directory-based module format.

But basically it's a style question.  What capitalization / style scheme should we use for the keys that are user-facing?  ie. `fsPath` vs `fs_path` vs. ???

And more generally, I think we are getting to the point we should be thinking about a MeTTa style guide so people know how to quickly interpret what kind of an object they have based on the capitalization / style scheme.

Reply:
>  I'll probably use Hjson or TOML (which do you prefer?) 

I didn't hear about Hjson yet and I used TOML only recently. Both formats are not as wide as JSON but both are human readable/writable and supported. I don't have a strong opinion here.

> What capitalization / style scheme should we use for the keys that are user-facing? ie. fsPath vs fs_path vs. ???

In MeTTa code I trying stick to Scheme code style (https://www.r6rs.org/final/html/r6rs/r6rs-Z-H-2.html#node_toc_node_sec_6.7) but I don't think it is the only option.

Reply:
Hjson is nice because most people think it's just JSON.  Basically it is the JSON that the javascript interpreter will accept.  As opposed to the oppressively obnoxiously strict JSON of the JSON spec.

Personally I like Hjson quite a bit.  But TOML would also be fine.

> I trying stick to Scheme code style (https://www.r6rs.org/final/html/r6rs/r6rs-Z-H-2.html#node_toc_node_sec_6.7)

'-' as a separator is fine with me.

I'm a little confused about capitalization recommendations however.  I'm not sure I'm reading it correctly, but is Scheme case-insensitive?  The document said said "The parameter names are case-insensitive: Vector is the same as vector." But I may have over-generalized that sentence.  Does Scheme have capitalization recommendations?

Reply:
> Does Scheme have capitalization recommendations?

I am not sure actually. But our atoms are case-sensitive and we can make our own suggestions.